opnode/rollup/driver/driver.go

```diff
@@ -2,71 +2,139 @@ package driver
+// keep making many sync steps if we can make sync progress
+const hot = time.Millisecond * 30
 
-type EngineDriver struct {
-	Log log.Logger
-	// API bindings to execution engine
-	RPC     DriverAPI
-	DL      Downloader
-	SyncRef rollupSync.SyncReference
+// check on sync regularly, but prioritize sync triggers with head updates etc.
+const cold = time.Second * 8
 
-	// The current driving force, to shutdown before closing the engine.
-	driveSub ethereum.Subscription
-	// There may only be 1 driver at a time
-	driveLock sync.Mutex
+// at least try every minute to sync, even if things are going well
+const max = time.Minute
 
-	EngineDriverState
+type Downloader interface {
+	Fetch(ctx context.Context, id eth.BlockID) (*types.Block, []*types.Receipt, error)
 }
 
-func (e *EngineDriver) Drive(ctx context.Context, l1Heads <-chan eth.HeadSignal) ethereum.Subscription {
-	e.driveLock.Lock()
-	defer e.driveLock.Unlock()
-	if e.driveSub != nil {
-		return e.driveSub
+type Driver struct {
+	log     log.Logger
+	rpc     DriverAPI
+	syncRef rollupSync.SyncReference
+	dl      Downloader
+	l1Heads <-chan eth.HeadSignal
+	done    chan chan error
+	genesis rollup.Genesis
+	state   // embedded engine state
+}
+
+func NewDriver(l2 DriverAPI, l1 l1.Source, log log.Logger, genesis rollup.Genesis) *Driver {
+	return &Driver{
+		log:     log,
+		rpc:     l2,
+		syncRef: rollupSync.SyncSource{L1: l1, L2: l2},
+		dl:      l1,
+		done:    make(chan chan error),
+		genesis: genesis,
+		state:   state{Genesis: genesis},
 	}
+}
 
-	e.driveSub = event.NewSubscription(NewDriverLoop(ctx, &e.EngineDriverState, e.Log, l1Heads, e))
-	return e.driveSub
+func (d *Driver) Start(ctx context.Context, l1Heads <-chan eth.HeadSignal) error {
+	d.l1Heads = l1Heads
+	if !d.requestUpdate(ctx, d.log, d) {
+		d.log.Error("failed to fetch engine head, defaulting to genesis")
+		d.updateHead(d.Genesis.L1, d.Genesis.L2)
+	}
+	go d.loop()
+	return nil
+}
+func (d *Driver) Close() error {
+	ec := make(chan error)
+	d.done <- ec
+	err := <-ec
+	return err
 }
 
-func (e *EngineDriver) requestEngineHead(ctx context.Context) (refL1 eth.BlockID, refL2 eth.BlockID, err error) {
-	refL1, refL2, _, err = e.SyncRef.RefByL2Num(ctx, nil, &e.Genesis)
-	return
+func (d *Driver) loop() {
+	ctx := context.Background()
+	backoff := cold
+	syncTicker := time.NewTicker(cold)
+	l2HeadPoll := time.NewTicker(time.Second * 14)
+
+	// exponential backoff, add 10% each step, up to max.
+	syncBackoff := func() {
+		backoff += backoff / 10
+		if backoff > max {
+			backoff = max
+		}
+		syncTicker.Reset(backoff)
+	}
+	syncQuickly := func() {
+		syncTicker.Reset(hot)
+		backoff = cold
+	}
+	onL2Update := func() {
+		// And we want to slow down requesting the L2 engine for its head (we just changed it ourselves)
+		// Request head if we don't successfully change it in the next 14 seconds.
+		l2HeadPoll.Reset(time.Second * 14)
+	}
+	defer syncTicker.Stop()
+	defer l2HeadPoll.Stop()
+
+	for {
+		select {
+		case <-d.done:
```

Comment:
are we not sending anything on the error channel that is passed through the done channel?

Reply:
Nope. I've switched some of the done channel to not return an error and I'll probably do that here. It doesn't feel like there is a well defined way to return an error only at close time.